lib/api/client-commands/captureNetworkCalls.js

```diff
@@ -0,0 +1,51 @@
+const ClientCommand = require('./_base-command.js');
+const {Logger} = require('../../utils');
+
+/**
+ * Capture outgoing network calls from the browser.
+ *
+ * @example
+ *  this.demoTest = function (browser) {
+ *    browser
+ *      .captureNetworkCalls((requestParams) => {
+ *        console.log('Request URL: ', requestParams.request.url);
+ *        console.log('Request method: ', requestParams.request.method);
+ *        console.log('Request headers: ', requestParams.request.headers);
+ *      })
+ *      .navigateTo('https://www.google.com');
+ *  };
+ *
+ * @method captureNetworkCalls
+ * @syntax .captureNetworkCalls(callback)
+ * @param {function} callback callback function to be called whenever a outgoing network call is made.
+ * @api protocol.userprompts
+ */
+class CaptureNetworkCalls extends ClientCommand {
+
+  performAction(callback) {
+
+    if (!this.api.isChrome()  && !this.api.isEdge()) {
+      const error =  new Error('CaptureNetworkCalls is not supported while using this driver');
+      Logger.error(error);
+
+      return callback(error);
+    }
+
+    const userCallback = this.userCallback;
+    
+    this.transportActions
+      .interceptNetworkCalls(userCallback, callback)
+      .catch(err => {
+        return err;
+      })
+      .then(result => callback(result));
```

Comment:
shouldn't it be userCallback instead of `callback` here? Considering `callback` is dummy callback the error would never appear in user provided callback

Reply:
Using `userCallback` here would pass the null result returned from method-mappings to the `userCallback`, which we don't want. `userCallback` is passed to the CDP command inside the definition specified in method-mappings, and is called whenever a new network call is made by CDP itself.

So, if some error occur, does `result` contain the information about that error?

Reply:
Yes I know the use case of userCallback but we should also present the error when the command fails. `result` will contain error as in catch we are returning error which would end up in `then` block

Reply:
So, would it be better to do something like this?
```js
  .then(result => {
    if (result instanceof Error) {
      userCallback(error);
    }
  });
```

Because we certainly don't want to send the returned `{value: null}` to `userCallback`.

Reply:
How can I test throwing error so that it ends up in `catch` block and then `then` block? I tried `throw Error('some error')` in method-mapping, but nothing happened. Then I tried passing `undefined` as callback to the command implementation in method-mapping, and the error got logged to the terminal and the test failed, but still no error in `catch` block.

<img width="931" alt="image" src="https://user-images.githubusercontent.com/39924567/171738414-c7a011ca-745b-40cd-ae77-fb5f932f2d39.png">

Reply:
Actually, we need a callback to be passed to the implementation in `method-mappings.js` which will then be used in the CDP command so that it can be called whenever there is a network request.

But if we pass just one callback, it is popped out from the args list (I need to see exactly where it happens) and is not passed to the `method-mappings.js`. That's why we need 2 callbacks here.

Also, the normal `callback` is called with the `result` value returned from the `method-mappings` implementation, which is `{value: null}` and should not be passed to the user-provided callback.

Reply:
One more thing is that, if we do like this:
```js
    this.transportActions
      .interceptNetworkCalls(userCallback, callback)
      .catch(err => {
        return err;
      })
      .then(result => callback(result));
```

The `callback` is called twice with the `result`, once here and the other time in the `performAction` method of the parent class [here](https://github.com/nightwatchjs/nightwatch/blob/b37a0feea8241b11388f4f6d33f0f96811cd91b5/lib/api/client-commands/_base-command.js#L29).

Reply:
Ok, I thought we are exposing both, but if only one callback is via the arguments, than it's fine.